src/imitation/algorithms/adversarial.py

```diff
@@ -147,65 +160,104 @@ def gen_policy(self) -> BaseRLModel:
     """Policy (i.e. the generator) being trained."""
     return self._gen_policy
 
-  def train_disc(self, n_steps=10, **kwargs):
+  def train_disc(self, total_timesteps: Optional[int] = None) -> None:
     """Trains the discriminator to minimize classification cross-entropy.
 
     Args:
-        n_steps (int): The number of training steps.
-        gen_obs (np.ndarray): See `_build_disc_feed_dict`.
-        gen_acts (np.ndarray): See `_build_disc_feed_dict`.
-        gen_next_obs (np.ndarray): See `_build_disc_feed_dict`.
+      total_timesteps: The number of transitions to sample from the generator
+        replay buffer and the expert demonstration dataset. (Half of the
+        samples are from each source).
+        By default, `self.disc_batch_size`. The number of optimizer updates
+        performed is `total_timesteps // self.n_disc_minibatch`.
+    """
+    if total_timesteps is None:
+      total_timesteps = self.disc_batch_size
+    n_epochs = total_timesteps // self.disc_batch_size
+    assert n_epochs >= 1, "should have at least one update"
+    for _ in range(n_epochs):
```

Comment:
Is there a difference between minibatches in different epochs? Seems like `self.disc_batch_size` doesn't really matter excepts as a default for `total_timesteps`.

Reply:
It's only used in two places: Here in `train_disc` (and therfore indirectly in `learn()`) and in `_build_feed_dict` for the default sample size.

Reply:
Looking over this code again I'm confused what's going on.

If `total_timesteps == self.disc_batch_size`, then we have 1 epoch of `self.n_disc_minibatch` minibatches of `self.disc_batch_size // self.n_disc_minibatch` size. So far, so good.

But if e.g. `total_timesteps == 2 * self.disc_batch_size`, then we have 2 epochs of `self.n_disc_minibatch` minibatches of `2 * self.disc_batch_size // self.n_disc_minibatch` size. This seems weird: the mini batch size should never change.

I'd scrap this epoch vs minibatch distinction. There's no meaningful difference here: you don't do anything at the start of each epoch. So there are three parameters (any two of which determine the third): batch size, minibatch size and number of minibatches. The question is which ones should the user provide as configuration parameters?

My inclination is to let the user specify batch size and minibatch size. Loosely, mini-batch size you want to set based on how noisy the gradient is, whereas batch size you want to set based on how much training time the discriminator should be able to get. So you can vary them somewhat independently. But I'd be OK with other choices too.

Reply:
Oops, yeah that should be `minibatch_size = self.disc_batch_size // self.n_disc_minibatch`

Reply:
Setting the minibatch size directly sounds good to me. (My original motivation was to mirror PPO2's hyperparameters)

Reply:
I think batch vs minibatch makes more sense in the case of PPO, where the *batch* is collecting a fresh dataset and the *minibatch* are gradient updates. But we're not doing any data collection here.

Reply:
Made a change that sets minibatch size directly rather than using `n_disc_minibatch`.